Fix connection string parsing for IPv6 host names.

Brar opened the pull request:
This was broken by changes introduced with the multiple hosts feature.

While at it, add a pretty pointless allocation optimization using
ReadonlySpan<char>.

Fixes #3802

vonzshik reviewed (approved):
LGTM

roji commented on src/Npgsql/MultiHostConnectorPool.cs:
```diff
@@ -26,11 +26,21 @@ public MultiHostConnectorPool(NpgsqlConnectionStringBuilder settings, string con
+                                .ToString()
+#endif
+                        );
+                        host = span.Slice(0, portSeparator).ToString();
```
Here there's no value in the Span, since you're getting a string out (so can just do shorter Substring if you want)

roji commented on src/Npgsql/NpgsqlConnection.cs:
```diff
@@ -180,19 +181,29 @@ void GetPoolAndSettings()
 
-                var portSeparator = settings.Host!.IndexOf(':');
-                if (!Path.IsPathRooted(settings.Host) && portSeparator != -1)
+                var portSeparator = settings.Host!.LastIndexOf(':');
```
Looks duplicated with MultiHostConnectorPool... Maybe DRY this (e.g. to a static internal method on PGUtil or NpgsqlConnectionStringBuilder)?

roji commented on test/Npgsql.Tests/ConnectionTests.cs:
```diff
@@ -528,6 +528,76 @@ public async Task TimezoneConnectionParam()
+                    throw new SuccessException("Multiple hosts in connection string is ignored for Multiplexing");
+                // We reset the cluster's state for multiple hosts
+                // Because other tests might have marked some of the hosts as disabled
+                ClusterStateCache.Clear();
```
Any thoughts on test parallelization here? There's a race condition where a host may get marked as disabled after this is called - at the very least maybe put a [NonParallelizable] on this test.

roji replied:
Sorry, missed it.

roji commented on test/Npgsql.Tests/MultipleHostsTests.cs:
```diff
@@ -518,6 +518,9 @@ public void Cluster_state_cache_basic()
             PoolManager.Reset();
+            // We reset the cluster's state for multiple hosts
+            // Because other tests might have marked some of the hosts as disabled
+            ClusterStateCache.Clear();
```
As above, we should consider test parallelization here /cc @vonzshik

roji replied:
Sorry, missed it.

roji reviewed (approved):
Looks good, see some comments.

Brar commented:
@roji, I've addressed the feedback regarding the actual Npgsql code changes but left the `ClusterStateCache` related things in the tests for @vonzshik since I'm by no means an expert in this area.

roji commented on src/Npgsql/Netstandard20/ReadOnlySpanOfCharExtensions.cs:
```diff
@@ -0,0 +1,18 @@
+    static class ReadOnlySpanOfCharExtensions
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int ParseInt(this ReadOnlySpan<char> span)
```
I'd personally not do this extension method over such a common type (which will appear in Intellisense) just for one call site...

roji replied:
Never mind, was already merged :) Maybe I'll clean up later.

Brar replied:
I'm pretty sure this will occur more often once we get used to using `ReadonlySpan<char>` for parsing things (as long as we have to stick with .NET Standard 2.0).

I guess the only reason why there isn't a shim already is the fact that you can't create a static extension method over Int32 (and perhaps because it sneaks in an allocation where you're actually trying to avoid it).

roji replied:
Yeah, it's all internal anyway, so doesn't matter so much either way.
